### ISSUE TITLE
Access to VSO Question

### ISSUE THREAD

**Issue by imsam67** [MicrosoftDocs/vsonline#97]
Is it safe to assume to that if a developer in a team is no longer in the team, simply removing his/her account from the repo will prevent any potential unauthorized access?

P.S. This is a question and not an issue but [the blog post](https://devblogs.microsoft.com/visualstudio/announcing-visual-studio-online-public-preview/) indicates "any feedback". If there's a better place for me to ask my questions, please let me know as I have a few more administration questions.

**Comment by jkeech** [comment]
This is a great place to ask questions. I want to make sure I understand the scenario correctly. Are you mostly concerned about source code access? If so, the behavior is essentially the same as if someone had cloned the repo locally.

When a user first creates an environment, they need to have access to the source code to clone the repo (either a public repo, or a private repo they currently have permissions to). After that, they will continue to need access to the repo in order to fetch updates or push any changes back. So if someone has a clone of your repo and they leave your team, they could still have access to the source code they previously cloned, but would not have the ability to see any future changes that you pushed, and they would not be able to push any of their changes back. All of this is determined by the backing source control provider though, such as GitHub. VSO itself doesn't control or manage source control permissions. For GitHub specifically, VSO uses an OAuth token for the user, so all interactions with GitHub are using the user's identity. If the user's permissions are revoked (or if they revoke the VSO OAuth app), then future push/pull operations will fail.

Hope that helps!

**Comment by imsam67** [comment]
@jkeech Thank you for your detailed response. 

The particular scenario is developing an app for a company using a distributed team of developers. The software is not open source and the company is pretty sensitive about protecting its intellectual property.

So, the most important two things we want to achieve are:

- Developers who are no longer in the team lose access to all the code as soon as they're terminated
- Developers can only work within the VSO environment and not be able to download the code into their local computers

VSO is pretty attractive at multiple levels for us and we're trying to figure out if we can achieve these two requirements with VSO.

Thanks again for your response.

**Comment by lostintangent** [comment]
@imsam67 As @jkeech mentioned, #1 would already be achieved by revoking the terminated employee's access from the version control system. Additionally, #2 would be semi-achieved once we allow you to share plans (#146), since we'd also allow you to revoke plan access, and therefore, environment access.

However, what we don't currently have a solution for is: preventing users to download files from the environment to their local machine. So it sounds like that might be a unique requirement from your perspective? Maybe we could re-purpose this issue to track that?

**Comment by imsam67** [comment]
@lostintangent Thank you for your response.

We can track that particular requirement separately as you wish. The general idea is that there are situations/projects that are pretty sensitive and project owners simply want to protect their intellectual property.

If the whole environment cannot be downloaded/exported, I feel we're 80% of the way there. Any project with a decent set of features is bound to have hundreds, if not thousands, of files and downloading them one by one is simply not practical. So, if this is the current set up with VSO, we may be able to use it for our purposes.

But ultimately, through some policies, project owners/managers should be able to control the environment fully and disable downloading/exporting of any files.

**Comment by jkeech** [comment]
@imsam67, @lostintangent, I think this requirement is a lot more involved than just being able to copy/export directly from the environment. Users could also exfiltrate data from the environment to another machine/service first. So I think a "locked-down" environment would also need to severely limit network access. #149 is probably applicable in this scenario if you run things in a private vnet that has tightly controlled network firewall/access policies.

**Comment by imsam67** [comment]
@jkeech @lostintangent I completely agree. There's one more but slightly less important issue which is "importing" files into the environment.

In our case, we've had projects where we didn't want developers to simply go out and grab an npm package or some plugin without running it through the proper procedure. Unfortunately, if we didn't have full control over the envrionment -- which we didn't -- some developer would still add them to the project without any regard for licensing or compatibility issues.

So, ideally, control should be both ways -- import and export. I think, as mentioned in #149 the idea of connecting to an Azure VNET makes sense.
